rfcs/RFC-think-tool.md

```diff
@@ -0,0 +1,301 @@
+- Feature Name: think-tool
+- Start Date: 2025-03-27
+
+# Summary
+
+[summary]: #summary
+
+Add a "think" tool that allows the model to reason through complex problems during response generation. This tool provides a dedicated space for the model to process information from tool call results, navigate complex decision trees, and improve the quality of responses in multi-step scenarios.
+
+# Motivation
+
+[motivation]: #motivation
+
+When handling complex tasks, the model often needs to reason through multiple steps, analyze information from previous tool calls, or plan a sequence of operations. Currently, this reasoning happens implicitly within the model's response generation, which can lead to:
+
+1. Incomplete reasoning due to token limitations
+2. Difficulty tracking state across multiple tool calls
+3. Mixing reasoning with user-facing content
+4. Reduced clarity in complex decision-making processes
+
+The "think" tool provides a dedicated space for this reasoning, allowing the model to:
+
+1. Process information from tool call results
+2. Navigate complex decision trees
+3. Plan multi-step operations
+4. Maintain state across interactions
+5. Improve the quality of responses in complex scenarios
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The "think" tool gives the model a space to work through complex problems step by step without showing this process to the user. Think of it as the model's scratch pad or internal monologue.
+
+When implementing features that use the "think" tool, you'll see code like this:
+
+```rust
+// Example of how a developer would use the think tool in their code
+let thought = "I need to analyze the file structure before suggesting changes";
+think_tool.invoke(thought);
+```
+
+For users, this feature is invisible by default but can be enabled:
+
+```shell
+# Enable the thinking feature
+q settings enable_thinking true
+```
```

Comment:
Can we make this more generic? I don't want to add a new setting for every tool.

Also potentially, we should have the ability to enable/disable any tool. Can we slot into a system like that?